src/mesh/python/x3d_generator.py

```diff
@@ -51,6 +59,65 @@
 elif args.mesh_type in ['vor_2d_mesh']:
     mesh = mesh_type_dict[args.mesh_type](bnd_per_dim, args.num_cells)
 
+# ------------------------------------------------------------------------------------------------ #
+# -- construct matid and region dictionary if region data arguments are available
+
+# -- set matid data (using region ids and bounds if available)
+matids = np.zeros(mesh.num_cells, dtype=int)
+reg_dict = {}
+reg_ids = []
+if (len(args.reg_ids) > 0):
+    # -- prepend reg_ids and reg_bnd_per_dim with universal region
+    reg_ids = args.reg_ids
+    reg_bnd_per_dim = args.reg_bnd_per_dim
+    reg_ids = [0] + reg_ids
+    reg_bnd_per_dim = args.bnd_per_dim + reg_bnd_per_dim
+    # -- initialize dictionary of cell lists
+    reg_dict = {reg_id: [] for reg_id in reg_ids}
+    nmats = len(reg_ids)
+    for cell in range(mesh.num_cells):
+        # -- initialize container for cell average coordinate
+        coord_av = np.array([0.0, 0.0, 0.0])
+        # -- average coordinates over faces first
+        for face in mesh.faces_per_cell[cell]:
+            # -- initialize container for face average coordinate
+            coord_face_av = np.array([0.0, 0.0, 0.0])
+            # -- get the node coordinates for this face to average
+            for node in mesh.nodes_per_face[face]:
+                for idim in range(mesh.ndim):
+                    coord_face_av[idim] += mesh.coordinates_per_node[node, idim]
+            # -- finish face average by dividing by number of nodes per face
+            coord_face_av /= mesh.num_nodes_per_face[face]
+            # -- add face average to cell average
+            coord_av += coord_face_av
+        # -- finish cell average by dividing by number of faces per cell
+        coord_av /= mesh.num_faces_per_cell[cell]
```

Comment:
Will this work for all concave cells?

Reply:
Good question - there are sort of a few answers to this.  The coordinate average here:
1.) can be outside of cell if the cell is concave,
2.) serves as an indicator for whether the cell is included in a particular region,
3.) but there region is merely an overly (e.g painting a label), so doesn't affect the mesh structure,
4.) cell can be included in a region without being entirely inside it, so there is error on the order of the cell resolution.

Reply:
Putting (1)-(4) together, if a cell is extremly concave, it is possible (pretty sure at least) most of it's volume will be out of the region, while the `coord_av` will be in the region.  This is why I had commented that this averaging should be a method in the `mesh_type` mesh classes (but it didn't seem urgent to do so).  Should I go ahead and make this change here (it is not too difficult)?

Reply:
> This is why I had commented that this averaging should be a method in the mesh_type mesh classes

Ah OK I misunderstood that comment above. 

I think this is fine as is for now (especially since, like you say, this is really a spatial error of a type that is already inherent in the mesh scheme), but what do you think about adding a comment somewhere here e.g. `TODO(RW) make coordinate averaging a separate method for each mesh_type to account for concave cells`? (That's 100 characters so you might have a better comment in mind)

Reply:
Just added a TODO comment for this.